### PR TITLE
Metric calculation as a separate class

### PR DESCRIPTION
**Pull request by rajasbansal**
- This allows separation of metric calculation from all tasks
- Users can define new additional metrics as classes which can be run every time the model runs
- AUROC calculation now does not return thresholds
- Thresholding logic removed from all tasks - users can request the values at specific thresholds using the output object, or agent.report can show only the thresholds which the user should care about
- Allow specifying custom metrics

### CODE REVIEW

File: src/autolabel/tasks/question_answering.py [hunk 1]
```diff
@@ -106,85 +100,24 @@ def get_explanation_prompt(self, example: Dict) -> str:
         )
 
     def eval(
-        self, llm_labels: List[LLMAnnotation], gt_labels: List[str]
+        self,
+        llm_labels: List[LLMAnnotation],
+        gt_labels: List[str],
+        additional_metrics: Optional[List[BaseMetric]] = [],
```
Comment (rajasbansal) [inline src/autolabel/tasks/question_answering.py:106]: lmk if passing additional metrics over a set of default metrics for a task makes sense - we could have this be a set of metrics so you can also remove some default metrics

File: src/autolabel/metrics/base.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from abc import ABC, abstractmethod
+from typing import List
+
+from autolabel.schema import LLMAnnotation, MetricResult
+
+
+class BaseMetric(ABC):
+    value: List[MetricResult]
```
[the remark this answers is not in the document]
Comment (nihit) [reply src/autolabel/metrics/base.py:8]: hmm, I think it's better to have this be stateless and return on compute if we're passing the inputs during the compute call - the intermediate way (inputs passed to compute, results stored in self.value) is a bit confusing?

**Review by nihit — COMMENTED** [review]
still need to finish reviewing but initial thoughts

File: src/autolabel/tasks/question_answering.py (hunk 1 above)
Comment (nihit) [reply src/autolabel/tasks/question_answering.py:106]: would a user specify additional metrics in the config ? (or, in the future world, initialize it in code themselves and pass it here) ?

Comment (rajasbansal) [reply src/autolabel/tasks/question_answering.py:106]: The hope is that these Metrics will be defined in code and the user can specify them in the agent.run command

Comment (rajasbansal) [reply src/autolabel/tasks/question_answering.py:106]: Added a test to show how a new metric can be added

**Review by nihit — APPROVED** [review]
lgtm!

**Pull request merged**